zcash_client_sqlite/src/wallet.rs

```diff
@@ -3088,13 +3089,82 @@ pub(crate) fn put_block(
     Ok(())
 }
 
+fn determine_fee(
+    conn: &rusqlite::Connection,
+    tx: &Transaction,
+) -> Result<Option<Zatoshis>, SqliteClientError> {
+    let value_balance: ZatBalance = tx
+        .sapling_bundle()
+        .map_or(ZatBalance::zero(), |s_bundle| *s_bundle.value_balance());
+
+    let value_balance = tx.orchard_bundle().map_or(Ok(value_balance), |o_bundle| {
+        (value_balance + *o_bundle.value_balance()).ok_or(BalanceError::Overflow)
+    })?;
+
+    match tx.transparent_bundle() {
+        Some(t_bundle) => {
+            let t_out_total = t_bundle
+                .vout
+                .iter()
+                .map(|o| o.value)
+                .sum::<Option<Zatoshis>>()
+                .ok_or(BalanceError::Overflow)?;
+            let value_balance = (value_balance - t_out_total).ok_or(BalanceError::Underflow)?;
+
+            if t_bundle.vin.is_empty() {
+                Zatoshis::try_from(value_balance)
+                    .map(Some)
+                    .map_err(SqliteClientError::from)
+            } else {
+                #[cfg(feature = "transparent-inputs")]
+                let mut result = Some(value_balance);
+
+                #[cfg(feature = "transparent-inputs")]
+                for t_in in &t_bundle.vin {
+                    match (
+                        result,
+                        get_wallet_transparent_output(conn, &t_in.prevout, true)?,
+                    ) {
+                        (Some(b), Some(out)) => {
+                            result = Some((b + out.txout().value).ok_or(BalanceError::Overflow)?)
+                        }
+                        _ => {
+                            result = None;
+                            break;
+                        }
+                    }
+                }
+
+                #[cfg(not(feature = "transparent-inputs"))]
+                let result = None;
+
+                Ok(result.map(Zatoshis::try_from).transpose()?)
+            }
+        }
+        None => {
+            // The sum of the value balances represents the total value of inputs not consumed by
+            // outputs. Without transparent outputs to consume any leftover value, the total is
+            // available for and treated as fee.
+            Ok(Some(
+                value_balance
+                    .try_into()
+                    .map_err(|_| BalanceError::Underflow)?,
+            ))
+        }
+    }
```

Comment:
This should all be using `TransactionData::fee_paid` (via the `Transaction` deref, so `tx.fee_paid(_)` should work). That method takes a `FnMut` closure that should therefore be able to do whatever it needs to with `conn` internally to look up the input outpoints.

Reply:
This would best be served via a breaking change to the signature of `TransactionData::fee_paid` to allow for the case where the output cannot be obtained by the closure. In retrospect, it's bad that that API makes it seem like that function is total; one has to then plumb the fact that the output is missing into the error type.

Reply:
Opened https://github.com/zcash/librustzcash/issues/1869 to track this; I think it's better to make this change in `main` after https://github.com/zcash/librustzcash/pull/1870 is merged.